service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt

```diff
@@ -0,0 +1,698 @@
+            }
+        }
+    } catch (e: Exception) {
+        logger.error("VardaUpdate: manual check needed: something went from while trying to add varda service need $serviceNeedId data: ${e.localizedMessage}")
+    }
+}
+
+fun retryUnsuccessfulServiceNeedVardaUpdates(db: Database.Connection, vardaClient: VardaClient) {
+    val unsuccessfullyUploadedServiceNeeds = db.read { it.getUnsuccessfullyUploadVardaServiceNeeds() }
+    unsuccessfullyUploadedServiceNeeds.forEach {
+        try {
+            if (it.existsInEvaka && it.childId != null)
+                handleUpdatedEvakaServiceNeed(db, vardaClient, it.evakaServiceNeedId)
+            else
+                handleDeletedEvakaServiceNeed(db, vardaClient, it.evakaServiceNeedId)
+        } catch (e: Exception) {
+            logger.error("VardaUpdate: got an error while trying to redo a failed service need varda update: ${e.localizedMessage}")
+        }
+    }
+}
+
+// Delete varda decision, placement and related varda fee decisions
+fun deleteServiceNeedDataFromVarda(vardaClient: VardaClient, vardaServiceNeed: VardaServiceNeed): List<String> {
+    val errors = mutableListOf<String>()
+    try {
+        errors.addAll(deleteServiceNeedAndRelatedDataFromVarda(vardaClient, vardaServiceNeed))
+    } catch (e: Exception) {
+        errors.add("VardaUpdate: error deleting varda service need ${vardaServiceNeed.evakaServiceNeedId} related data: ${e.localizedMessage}")
+    }
+    return errors
+}
+
+// Add child if missing, service need, placement and mandatory fee decision(s)
+fun addServiceNeedDataToVarda(db: Database.Connection, vardaClient: VardaClient, evakaServiceNeed: EvakaServiceNeedInfoForVarda, newVardaServiceNeed: VardaServiceNeed): List<String> {
+    val errors = mutableListOf<String>()
+
+    try {
+        // Todo: "nettopalvelu"-unit children do not have fee data
+        val serviceNeedFeeData = db.read { it.getServiceNeedFeeData(null, evakaServiceNeed.id) }.firstOrNull()
+        if (serviceNeedFeeData != null && (serviceNeedFeeData.feeDecisionIds.isNotEmpty() || serviceNeedFeeData.voucherValueDecisionIds.isNotEmpty())) {
+            if (evakaServiceNeed.ophOrganizerOid.isNullOrEmpty()) throw Exception("VardaUpdate: service need ${evakaServiceNeed.id} related oph_orginizer_id is null or empty")
+
+            newVardaServiceNeed.vardaChildId = getOrCreateVardaChildByOrganizer(db, vardaClient, evakaServiceNeed.childId, evakaServiceNeed.ophOrganizerOid, vardaClient.sourceSystem)
+            newVardaServiceNeed.vardaDecisionId = createDecisionToVarda(vardaClient, newVardaServiceNeed.vardaChildId, evakaServiceNeed)
+            newVardaServiceNeed.vardaPlacementId = createPlacementToVarda(vardaClient, newVardaServiceNeed.vardaDecisionId!!, evakaServiceNeed)
+            newVardaServiceNeed.vardaFeeDataIds = serviceNeedFeeData.feeDecisionIds.map { fdId ->
+                val feeDecision = db.read { it.getFeeDecisionsByIds(listOf(fdId)) }.first()
```

Comment:
I wonder if it should be `createFeeDataToVardaFromFeeDecision` that fetches the feeDecision details it needs. Somehow it seems a little heavy to do the db.read here instead of just passing the id forward.